bfvmm/integration/arch/intel_x64/test_all_inheritance.cpp

```diff
@@ -0,0 +1,98 @@
+//
+// Copyright (C) 2019 Assured Information Security, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// TIDY_EXCLUSION=-cert-err58-cpp
+//
+// Reason:
+//     This test triggers on the use of a std::mutex being globally defined
+//     from the EPT map.
+//
+
+#include <vmm.h>
+
+// -----------------------------------------------------------------------------
+// vCPU
+// -----------------------------------------------------------------------------
+
+ept::mmap g_guest_map;
+
+void
+init()
+{
+    ept::identity_map(g_guest_map, MAX_PHYS_ADDR);
+}
+
+namespace test
+{
+
+class vcpu : public bfvmm::intel_x64::vcpu
```

Comment:
The important thing to notice here is that the original vCPU extension model still works. In fact, all of the unit tests and integration tests still work fine. The only change that will be needed by downstream users is to remove the gsl::not_null which includes Boxy which also still works fine.

Reply:
Oh, and near the top we use the init() so using a hybrid approach also works with this design. I could have implemented this without any changes to entry.cpp (that is what I had working last night before bed), but I wanted the old style to work with the new style and the best way to do that was to move the once_call to the entry. The logic is identical so we are not introducing possible issues with threading. All this does is move the once_call from the vcpu constructor to after the vcpu is fully constructed.